protocolupgrade/engine.go

```diff
@@ -0,0 +1,252 @@
+	"github.com/cenkalti/backoff"
+	"github.com/golang/protobuf/proto"
+)
+
+type protocolUpgradeProposal struct {
+	blockHeight        uint64
+	vegaReleaseTag     string
+	dataNodeReleaseTag string
+	accepted           map[string]struct{}
+}
+
+func protocolUpgradeProposalID(upgradeBlockHeight uint64, vegaReleaseTag string, dataNodeReleaseTag string) string {
+	return fmt.Sprintf("%v@%v/%v", vegaReleaseTag, upgradeBlockHeight, dataNodeReleaseTag)
+}
+
+func (p *protocolUpgradeProposal) approvers() []string {
+	accepted := make([]string, 0, len(p.accepted))
+	for k := range p.accepted {
+		accepted = append(accepted, k)
+	}
+	sort.Strings(accepted)
+	return accepted
+}
+
+type ValidatorTopology interface {
+	IsTendermintValidator(pubkey string) bool
+	IsSelfTendermintValidator() bool
+	NumberOfTendermintValidators() uint
+}
+
+type Commander interface {
+	CommandSync(ctx context.Context, cmd txn.Command, payload proto.Message, f func(error), bo *backoff.ExponentialBackOff)
+}
+
+type Broker interface {
+	Send(event events.Event)
+}
+
+type Engine struct {
+	log      *logging.Logger
+	config   Config
+	broker   Broker
+	topology ValidatorTopology
+	hashKeys []string
+
+	currentBlockHeight uint64
+	activeProposals    map[string]*protocolUpgradeProposal
+	events             map[string]*eventspb.ProtocolUpgradeEvent
+	lock               sync.RWMutex
+
+	requiredMajority num.Decimal
+	upgradeStatus    *types.UpgradeStatus
+}
+
+func New(log *logging.Logger, config Config, broker Broker, topology ValidatorTopology) *Engine {
+	log = log.Named(namedLogger)
+	log.SetLevel(config.Level.Get())
+	return &Engine{
+		activeProposals: map[string]*protocolUpgradeProposal{},
+		events:          map[string]*eventspb.ProtocolUpgradeEvent{},
+		log:             log,
+		config:          config,
+		broker:          broker,
+		topology:        topology,
+		hashKeys:        []string{(&types.PayloadProtocolUpgradeProposals{}).Key()},
+		upgradeStatus:   &types.UpgradeStatus{},
+	}
+}
+
+func (e *Engine) OnRequiredMajorityChanged(_ context.Context, requiredMajority num.Decimal) error {
+	e.requiredMajority = requiredMajority
+	return nil
+}
+
+// UpgradeProposal records the intention of a validator to upgrade the protocol to a release tag at block height.
+func (e *Engine) UpgradeProposal(ctx context.Context, pk string, upgradeBlockHeight uint64, vegaReleaseTag, dataNodeReleaseTag string) error {
+	e.lock.RLock()
+	defer e.lock.RUnlock()
+	if !e.topology.IsTendermintValidator(pk) {
+		// not a tendermint validator, so we don't care about their intention
+		return nil
+	}
+
+	if upgradeBlockHeight <= e.currentBlockHeight {
+		return errors.New("upgrade block earlier than current block height")
+	}
+
+	if e.upgradeStatus.AcceptedReleaseInfo != nil {
+		return errors.New("protocol upgrade already scheduled")
+	}
+
+	if !e.topology.IsSelfTendermintValidator() {
+		return errors.New("validator is not a tendermint validator - cannot decide on upgrade")
+	}
```

Comment:
Not sure we should stop here if we are not a validator ourselves right? Even non-validators needs to process the rest of the state after that?

Reply:
removed